SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp

```diff
@@ -0,0 +1,278 @@
+	// Singleton : 1
+	// Grid : 2 - 5
+	// Singleton: 6
+	VirtualWorkerId NextWorkerIdToAssign = FirstVirtualWorkerId;
+	for (auto& Elem : LayerNameToLBStrategy)
+	{
+		UAbstractLBStrategy* LBStrategy = Elem.Value;
+		VirtualWorkerId MinimumRequiredWorkers = LBStrategy->GetMinimumRequiredWorkers();
+
+		VirtualWorkerId LastVirtualWorkerIdToAssign = NextWorkerIdToAssign + MinimumRequiredWorkers - 1;
+		if (LastVirtualWorkerIdToAssign > LastVirtualWorkerId)
+		{
+			UE_LOG(LogLayeredLBStrategy, Error, TEXT("LayeredLBStrategy was not given enough VirtualWorkerIds to meet the demands of the layer strategies."));
+			return;
+		}
+		UE_LOG(LogLayeredLBStrategy, Log, TEXT("LayeredLBStrategy assigning VirtualWorkerIds %d to %d to Layer %s"), NextWorkerIdToAssign, LastVirtualWorkerIdToAssign, *Elem.Key.ToString());
+		LBStrategy->SetVirtualWorkerIds(NextWorkerIdToAssign, LastVirtualWorkerIdToAssign);
+
+		for (VirtualWorkerId id = NextWorkerIdToAssign; id <= LastVirtualWorkerIdToAssign; id++)
+		{
+			VirtualWorkerIdToLayerName.Add(id, Elem.Key);
+		}
+
+		NextWorkerIdToAssign += MinimumRequiredWorkers;
+	}
+}
+
+FName ULayeredLBStrategy::GetLayerNameForClass(const TSubclassOf<AActor> Class) const
+{
+	if (Class == nullptr)
+	{
+		return NAME_None;
+	}
+
+	UClass* FoundClass = Class;
+	TSoftClassPtr<AActor> ClassPtr = TSoftClassPtr<AActor>(FoundClass);
+
+	while (FoundClass != nullptr && FoundClass->IsChildOf(AActor::StaticClass()))
+	{
+		if (const FName* Layer = ClassPathToLayer.Find(ClassPtr))
+		{
+			FName LayerHolder = *Layer;
+			if (FoundClass != Class)
+			{
+				ClassPathToLayer.Add(TSoftClassPtr<AActor>(Class), LayerHolder);
+			}
+			return LayerHolder;
+		}
+
+		FoundClass = FoundClass->GetSuperClass();
+		ClassPtr = TSoftClassPtr<AActor>(FoundClass);
+	}
+
+	// No mapping found so set and return default actor group.
+	ClassPathToLayer.Add(TSoftClassPtr<AActor>(Class), SpatialConstants::DefaultLayer);
+	return SpatialConstants::DefaultLayer;
+}
+
+bool ULayeredLBStrategy::IsSameWorkerType(const AActor* ActorA, const AActor* ActorB) const
+{
+	if (ActorA == nullptr || ActorB == nullptr)
+	{
+		return false;
+	}
+
+	const FName& LayerA = GetLayerNameForClass(ActorA->GetClass());
+	const FName& LayerB = GetLayerNameForClass(ActorB->GetClass());
+
+	return (LayerA == LayerB);
+}
+
+// Note: this is returning whether this is one of the workers which can simulate the layer. If there are
+// multiple workers simulating the layer, there's no concept of owner. This is left over from the way
+// ActorGroups could own entities, and will be removed in the future.
+bool ULayeredLBStrategy::IsLayerOwner(const FName Layer) const
+{
+	return *VirtualWorkerIdToLayerName.Find(LocalVirtualWorkerId) == Layer;
+}
+
+FName ULayeredLBStrategy::GetLayerNameForActor(const AActor& Actor) const
+{
+	return GetLayerNameForClass(Actor.GetClass());
+}
+
+void ULayeredLBStrategy::AddStrategyForLayer(const FName LayerName, UAbstractLBStrategy* LBStrategy)
```

Comment:
If LayerName is passed by ref that means we only copy once instead of twice? (assuming there will always be a copy when adding to map)

Reply:
yup, this is what happens when I'm trying to turn around comments in 10 minutes before a meeting. oops.  :)